{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d279aa0d",
   "metadata": {},
   "source": [
    "# Using the data from prev scRNAseq datasets to do probe selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b10acc53",
   "metadata": {},
   "source": [
    "Start by loading prev scRNAseq datasets into the spapros framework using [this tutorial](https://spapros.readthedocs.io/en/latest/_tutorials/spapros_tutorial_basic_selection.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b9897f1",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c115d19",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import spapros as sp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from pathlib import Path\n",
    "import iss_analysis.io as io\n",
    "import iss_analysis.pick_genes as pick\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "from scipy.stats import entropy\n",
    "import utils as ut\n",
    "\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dde5405b",
   "metadata": {},
   "source": [
    "## Load reference train dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3afecb0",
   "metadata": {},
   "source": [
    "We take the raw count data that we generated in the MOs panel notebook. For this, we use the tools we built to use Petr's panel (a bit bad that we need to import the whole iss_analysis for using these simple functions). We need to filter for the main classes of neurons we care about, plus region of interest and neurotransmitters. \n",
    "\n",
    "Without the filtering by class, we get:\n",
    "\n",
    "| Code | Cell type            |\n",
    "|------|----------------------|\n",
    "| 01   | IT-ET Glut           |\n",
    "| 02   | NP-CT-L6b Glut       |\n",
    "| 03   | OB-CR Glut           |\n",
    "| 05   | OB-IMN GABA          |\n",
    "| 06   | CTX-CGE GABA         |\n",
    "| 07   | CTX-MGE GABA         |\n",
    "| 08   | CNU-MGE GABA         |\n",
    "| 09   | CNU-LGE GABA         |\n",
    "| 30   | Astro-Epen           |\n",
    "| 32   | OEC                  |\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c737b579",
   "metadata": {},
   "outputs": [],
   "source": [
    "download_base = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel')\n",
    "abc_cache = AbcProjectCache.from_cache_dir(download_base)\n",
    "\n",
    "abc_cache.current_manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14b2bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "reference = io.main_yao_2023(abc_cache, \n",
    "                  'WMB-10Xv3', \n",
    "                  ['WMB-10Xv3-Isocortex-2', 'WMB-10Xv3-Isocortex-1'], \n",
    "                  download_base,\n",
    "                  taxa_subclass = [\n",
    "                      \n",
    "                    \"001 CLA-EPd-CTX Car3 Glut\",\n",
    "                    \"002 IT EP-CLA Glut\",\n",
    "                    \"003 L5/6 IT TPE-ENT Glut\",\n",
    "                    \"004 L6 IT CTX Glut\",\n",
    "                    \"005 L5 IT CTX Glut\",\n",
    "                    \"006 L4/5 IT CTX Glut\",\n",
    "                    \"007 L2/3 IT CTX Glut\",\n",
    "                    \"020 L2/3 IT RSP Glut\",\n",
    "                    \"022 L5 ET CTX Glut\",\n",
    "                    \"029 L6b CTX Glut\",\n",
    "                    \"030 L6 CT CTX Glut\",\n",
    "                    \"032 L5 NP CTX Glut\",\n",
    "                    \"046 Vip Gaba\",\n",
    "                    \"049 Lamp5 Gaba\",\n",
    "                    \"050 Lamp5 Lhx6 Gaba\",\n",
    "                    \"047 Sncg Gaba\",\n",
    "                    \"053 Sst Gaba\",\n",
    "                    \"056 Sst Chodl Gaba\",\n",
    "                    \"051 Pvalb chandelier Gaba\",\n",
    "                    \"052 Pvalb Gaba\"\n",
    "                    ],\n",
    "                  region_of_interest = 'MO-FRP', \n",
    "                  neurotransmitters = ['GABA', 'Glut'], \n",
    "                  extract_csv = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7090f7b2",
   "metadata": {},
   "source": [
    "And now we attempt to run the software with predicting subclass as a target. There is very little CNU-LGE GABA in the isocortex as per Yao 2023, but hey"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53593c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(reference.obs['class'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b16c9e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(reference.obs['subclass'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a35a509a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count observations per subclass\n",
    "subclass_counts = reference.obs['subclass'].value_counts().sort_values(ascending=False)\n",
    "\n",
    "# Make a dataframe\n",
    "df_subclass = pd.DataFrame({\n",
    "    \"subclass\": subclass_counts.index,\n",
    "    \"count\": subclass_counts.values\n",
    "})\n",
    "\n",
    "# Add a row for the total\n",
    "df_subclass.loc[len(df_subclass)] = [\"TOTAL\", df_subclass[\"count\"].sum()]\n",
    "\n",
    "# Show the dataframe\n",
    "print(df_subclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5717e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# total reads per transcript\n",
    "reads_per_transcript = np.array(reference.X.sum(axis=0)).flatten()\n",
    "\n",
    "df_transcripts = pd.DataFrame({\n",
    "    \"gene\": reference.var_names,\n",
    "    \"total_reads\": reads_per_transcript\n",
    "}).sort_values(\"total_reads\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59200aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(df_transcripts[\"total_reads\"], bins=200, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e35f3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# total reads per cell\n",
    "reads_per_cell = np.array(reference.X.sum(axis=1))\n",
    "\n",
    "reference.obs[\"total_reads\"] = reads_per_cell\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa98c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference.obs.total_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec91d759",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(reference.obs.total_reads, bins=200, log=True)\n",
    "plt.xlabel(\"Reads per cell\")\n",
    "plt.ylabel(\"Number of cells (log scale)\")\n",
    "plt.title(\"Distribution of reads per cell\")\n",
    "\n",
    "# Force scientific offset notation (1, 2, 3, ... × 1e6)\n",
    "formatter = plt.ScalarFormatter(useMathText=True)\n",
    "formatter.set_scientific(True)\n",
    "formatter.set_powerlimits((0, 0))  # always show offset\n",
    "plt.gca().xaxis.set_major_formatter(formatter)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff2f36d",
   "metadata": {},
   "source": [
    "And we process it like they do. Select only the first 100 genes for efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0158e468",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(reference)\n",
    "sc.pp.log1p(reference)\n",
    "sc.pp.highly_variable_genes(reference,flavor=\"cell_ranger\",n_top_genes=1000)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e800911",
   "metadata": {},
   "source": [
    "## Run selector"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "260d7a42",
   "metadata": {},
   "source": [
    "Run vanilla selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40d40930",
   "metadata": {},
   "outputs": [],
   "source": [
    "selector = sp.se.ProbesetSelector(reference, n=100, celltype_key=\"subclass\", verbosity=1, n_jobs=-1, save_dir=\"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/spapros\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0c44ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "selector.select_probeset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a68a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "selector.probeset[selector.probeset[\"selection\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42964ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "lookup = pd.DataFrame({\n",
    "    \"gene_symbol\": reference.var['gene_symbol'],\n",
    "    \"feature_id\": reference.var.index\n",
    "})\n",
    "\n",
    "selected = selector.probeset[selector.probeset[\"selection\"]].copy()\n",
    "\n",
    "selected = selected.merge(lookup, \n",
    "                          left_index=True,\n",
    "                          right_on=\"feature_id\",\n",
    "                          how=\"left\")\n",
    "\n",
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da17090c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = 100  # control rows\n",
    "pd.options.display.max_columns = 10  # control cols"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe1a4c4c",
   "metadata": {},
   "source": [
    "## Compare to the greedy algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61570f6e",
   "metadata": {},
   "source": [
    "We need to take this probeset and compare it to the greedy algorithm using the spapros metrics and, if need be, the Znamenskiy metrics. Let's run it again with the new filters. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "441293e6",
   "metadata": {},
   "source": [
    "df, gene_names = io.main_yao_2023(abc_cache, \n",
    "                  'WMB-10Xv3', \n",
    "                  ['WMB-10Xv3-Isocortex-2', 'WMB-10Xv3-Isocortex-1'], \n",
    "                  download_base,\n",
    "                  taxa_class = ['01 IT-ET Glut', '06 CTX-CGE GABA', '07 CTX-MGE GABA',  '08 CNU-MGE GABA', '09 CNU-LGE GABA'],\n",
    "                  region_of_interest = 'MO-FRP', \n",
    "                  neurotransmitters = ['GABA', 'Glut'], \n",
    "                  extract_csv = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac84ec45",
   "metadata": {},
   "outputs": [],
   "source": [
    "pick.main(\n",
    "    '/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/greedy_panels',\n",
    "    efficiency=0.01,\n",
    "    datapath='/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel',\n",
    "    subsample=1,\n",
    "    classify=\"subclass\",\n",
    "    dataset = 'yao_2023'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b633d65e",
   "metadata": {},
   "source": [
    "## Evaluate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8096f230",
   "metadata": {},
   "source": [
    "We now have two alternatives. Now, we need to load them both as probesets in the spapros framework and compare. Ideally, we would also build the DE-only pipeline. \n",
    "\n",
    "In order to do this, we need:\n",
    "\n",
    "- The training dataset loaded (10xv3)\n",
    "- The two probesets loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ba1eef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "selector = sp.se.ProbesetSelector(reference, n=100, celltype_key=\"subclass\", verbosity=1, n_jobs=-1, save_dir=\"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/spapros\")\n",
    "lookup = pd.DataFrame({\n",
    "    \"gene_symbol\": reference.var['gene_symbol'],\n",
    "    \"feature_id\": reference.var.index\n",
    "})\n",
    "\n",
    "selected = selector.probeset[selector.probeset[\"selection\"]].copy()\n",
    "\n",
    "selected = selected.merge(lookup, \n",
    "                          left_index=True,\n",
    "                          right_on=\"feature_id\",\n",
    "                          how=\"left\")\n",
    "\n",
    "spapros_panel = list(selected.index)\n",
    "\n",
    "spapros_panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "465d8f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "znam = np.load('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/greedy_panels/greedy_panelsgenes_subclass_e0.01_s1_20250919_165916.npz', allow_pickle=True)\n",
    "panel = list(znam['gene_names'][znam['include_genes'] == True])\n",
    "panel = pd.DataFrame(panel)\n",
    "\n",
    "lookup = pd.DataFrame({\n",
    "    \"gene_symbol\": reference.var['gene_symbol'],\n",
    "    \"feature_id\": reference.var.index\n",
    "})\n",
    "\n",
    "# Build dict: gene_symbol → list of feature_ids\n",
    "#SOME GENE SYMBOLS HAVE MORE THAN ONE EMSEBBL ENTRY BUT NOT OURS\n",
    "multi_mapping = lookup.groupby(\"gene_symbol\")[\"feature_id\"].apply(list).to_dict()\n",
    "\n",
    "# Map panel values, may return lists\n",
    "panel_feature_ids = panel[0].map(multi_mapping)\n",
    "\n",
    "petr_panel = list(panel_feature_ids)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f239af",
   "metadata": {},
   "outputs": [],
   "source": [
    "petr_str = [str(x[0]) for x in petr_panel]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8db67ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "spapros_panel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5125398",
   "metadata": {},
   "source": [
    "And compare:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104b92dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_sets = sp.se.select_reference_probesets(reference, n = 100, obs_key = \"subclass\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b9fe3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluator = sp.ev.ProbesetEvaluator(reference, verbosity=2, celltype_key = 'subclass', results_dir=\"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/evaluation\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af1fb429",
   "metadata": {},
   "source": [
    "This step is actually kinda long, maybe because I have no gpu?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4404706",
   "metadata": {},
   "outputs": [],
   "source": [
    "for set_id, df in reference_sets.items():\n",
    "    gene_set = df[df[\"selection\"]].index.to_list()\n",
    "    evaluator.evaluate_probeset(gene_set, set_id=set_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd37f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "set_id = \"greedy_panel\"\n",
    "evaluator.evaluate_probeset(petr_panel, set_id=set_id)\n",
    "set_id = \"spapros_panel\"\n",
    "evaluator.evaluate_probeset(spapros_panel, set_id=set_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9867ff1",
   "metadata": {},
   "source": [
    "## Evaluate in test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61b41cfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_test = io.main_yao_2023(abc_cache, \n",
    "                  'WMB-10Xv2', \n",
    "                  ['WMB-10Xv2-Isocortex-3', 'WMB-10Xv2-Isocortex-1', 'WMB-10Xv2-Isocortex-4', 'WMB-10Xv2-Isocortex-2'], \n",
    "                  download_base,\n",
    "                  taxa_class = ['01 IT-ET Glut', '06 CTX-CGE GABA', '07 CTX-MGE GABA',  '08 CNU-MGE GABA', '09 CNU-LGE GABA'],\n",
    "                  region_of_interest = 'MO-FRP', \n",
    "                  neurotransmitters = ['GABA', 'Glut'], \n",
    "                  extract_csv = False)\n",
    "\n",
    "sc.pp.filter_genes(reference_test, min_cells=1)\n",
    "sc.pp.normalize_total(reference_test)\n",
    "sc.pp.log1p(reference_test)\n",
    "sc.pp.highly_variable_genes(reference_test,flavor=\"cell_ranger\",n_top_genes=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a696848f",
   "metadata": {},
   "source": [
    "## Running Petr evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7c5a46b",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_df, gene_names = io.read_yao_2023(\"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/\")\n",
    "train_set, tesset, cluster_labels = pick.train_test_split(exons_df, classify_by = \"subclass\", gene_filter=r\"\\d\", efficiency=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b5d29a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "znam = np.load('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/greedy_panels/greedy_panelsgenes_subclass_e0.01_s1_20250919_165916.npz', allow_pickle=True)\n",
    "gene_set = list(znam['gene_names'][znam['include_genes'] == True])\n",
    "gene_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f624601",
   "metadata": {},
   "outputs": [],
   "source": [
    "accuracy_train, accuracy_test = pick.evaluate_gene_set(train_set, \n",
    "    tesset, \n",
    "    gene_set, \n",
    "    gene_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "054fbfb5",
   "metadata": {},
   "source": [
    "## Running expression constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2764e4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading the yao2021 reference\n",
    "\n",
    "ALM_subclasses = [\n",
    "    \"Vip\",\n",
    "    \"Lamp5\",\n",
    "    \"Scng\",\n",
    "    \"Sst Chodl\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"L2/3 IT CTX\",\n",
    "    \"L4/5 IT CTX\",\n",
    "    \"L5 IT CTX\",\n",
    "    \"L6 IT CTX\",\n",
    "    \"L5 PT CTX\",\n",
    "    \"L4 RSP-ACA\",\n",
    "    \"L5/6 NP CTX\",\n",
    "    \"L6 CT CTX\",\n",
    "    \"L6b CTX\",\n",
    "]\n",
    "\n",
    "\n",
    "datapath = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/yao_2021')\n",
    "\n",
    "reference = io.load_yao_2021_to_anndata(datapath, 'ALM', ALM_subclasses)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3986ced",
   "metadata": {},
   "source": [
    "To compare the expression thresholds on our expression matrix to the expression thresholds on the Allen matrix, we need to have them on the same units. The average expression on the Allen dataset is the trimmed mean (percentiles 25-75) of the log2(CPM+1), where CPM is a normalised counts per cell *10e6. In the training set for spapros, data is normalised per cell to add up to the median expression level in the dataset, and then taken as loge(data), where loge is the natural logarithm.\n",
    "\n",
    "$ allen_{i, j} = \\log_2(\\frac{read_{i, j}}{read_j}*10^6+1) $"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5918c0",
   "metadata": {},
   "source": [
    "Therefore:\n",
    "\n",
    "$ \\frac{read_{i, j}}{read_j} = \\frac{2^{\\text{allen}_{i,j}} - 1}{10^6} $\n",
    "\n",
    "\n",
    "Where $read_{i, j}$ are the reads for gene i in cell j, and $read_j$ are the total reads in cell j\n",
    "\n",
    "We want to choose two expression constraints in this space, so we will chose 4 as a low constrain (Petr said it once) and 12 as a high constraint (Sst, which we can barely see)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ccb2bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "low_threshold_allen = 4\n",
    "high_threshold_allen = 11"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3df7973",
   "metadata": {},
   "source": [
    "And we make them into thresholds for spapros\n",
    "\n",
    "$ spapros_{i, j} = \\log_{e}(\\frac{read_{i, j}}{read_j} * median +1) $"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1932184d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def allen_to_spapros(allen, spapros_median):\n",
    "    # get the ratio per cell\n",
    "    ratio = (np.power( 2, allen) - 1)/10e6\n",
    "\n",
    "    #turn it into spapros\n",
    "    spapros = np.log1p(ratio*spapros_median) #what a funny little function\n",
    "\n",
    "    return spapros\n",
    "\n",
    "def get_median_expression(adata):\n",
    "    # Add counts up\n",
    "    cell_sums = np.array(adata.X.sum(axis=0)).flatten()\n",
    "    plt.hist(cell_sums, bins = 200, log=True)\n",
    "    plt.title('Total gene counts per cell')\n",
    "    plt.xlabel('Total gene counts')\n",
    "    plt.ylabel('#cells')\n",
    "    median = np.median(cell_sums)\n",
    "\n",
    "    return median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b48ca935",
   "metadata": {},
   "outputs": [],
   "source": [
    "median = get_median_expression(reference)\n",
    "print(f'median cell expression is {median}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72d0af72",
   "metadata": {},
   "outputs": [],
   "source": [
    "low_threshold_spapros = allen_to_spapros(low_threshold_allen, median)\n",
    "high_threshold_spapros = allen_to_spapros(high_threshold_allen, median)\n",
    "\n",
    "print(f'{low_threshold_spapros}, {high_threshold_spapros}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9410a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_filtered = reference.copy()\n",
    "\n",
    "sc.pp.filter_genes(reference_filtered, min_cells=1)\n",
    "sc.pp.normalize_total(reference_filtered)\n",
    "sc.pp.log1p(reference_filtered)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b55419",
   "metadata": {},
   "outputs": [],
   "source": [
    "median = get_median_expression(reference_filtered)\n",
    "print(f'median log1p expression is {median}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6504db5",
   "metadata": {},
   "source": [
    "The percentiles are probably screwing w us. Maybe just look at the number of counts of somatostatin in the somatostatin cluster?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1781cb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_idx = reference.var_names.get_loc('Sst')\n",
    "print(gene_idx)\n",
    "gene_expression = reference.X[:,gene_idx].toarray().flatten()\n",
    "\n",
    "sst_mask = reference.obs['subclass_label'] == 'Sst'\n",
    "print(sum(sst_mask))\n",
    "gene_expression = gene_expression[sst_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36e37e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_idx = reference.var_names.get_loc('Sst')\n",
    "print(gene_idx)\n",
    "gene_expression = reference.X[:,gene_idx].toarray().flatten()\n",
    "print(len(gene_expression))\n",
    "\n",
    "#sst_mask = reference.obs['subclass_label'] == 'Sst'\n",
    "#print(sum(sst_mask))\n",
    "#gene_expression = gene_expression[sst_mask]\n",
    "\n",
    "plt.hist(gene_expression, bins = 200, log = True)\n",
    "plt.title('Expression levels of Sst globally (counts)')\n",
    "print(np.percentile(gene_expression, 99))\n",
    "print(len(gene_expression[gene_expression>np.percentile(gene_expression, 99)]))\n",
    "[p_25, p_75] = [np.percentile(gene_expression, q) for q in [25, 75]]\n",
    "print([p_25, p_75])\n",
    "for q in [p_25, p_75]:\n",
    "    plt.axvline(q, color = 'red') \n",
    "plt.axvline(np.percentile(gene_expression, 99), color='green')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fd3b3fd",
   "metadata": {},
   "source": [
    "Maybe helpful?\n",
    "\n",
    "> EDIT: I was able to get closer to the values in `trimmed_means.csv` by taking the `log2(CPM + 1)` of all cells, then excluding the top and bottom 25%, and taking the mean of the remaining values. As an example, for a single cluster and a single gene, the `trimmed_means.csv` value is **6.651905**, and I’m getting **6.652420762158425**.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06676309",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4a0233a",
   "metadata": {},
   "outputs": [],
   "source": [
    "trimmed_expression = gene_expression[(gene_expression>p_25)&(gene_expression<p_75)]\n",
    "trimmed_mean = np.mean(trimmed_expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4884685f",
   "metadata": {},
   "outputs": [],
   "source": [
    "trimmed_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c465f410",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After we normalise, the trimmed mean expression of Sst in Sst neurons: \n",
    "\n",
    "gene_idx = reference_filtered.var_names.get_loc('Sst')\n",
    "print(gene_idx)\n",
    "gene_expression = reference_filtered.X[:,gene_idx].toarray().flatten()\n",
    "\n",
    "sst_mask = reference_filtered.obs['subclass_label'] == 'Sst'\n",
    "print(sum(sst_mask))\n",
    "gene_expression = gene_expression[sst_mask]\n",
    "\n",
    "plt.hist(gene_expression, bins = 200, log = True)\n",
    "plt.title('Expression levels of Sst in Sst cells (counts)')\n",
    "[p_25, p_75] = [np.percentile(gene_expression, q) for q in [25, 75]]\n",
    "print([p_25, p_75])\n",
    "for q in [p_25, p_75]:\n",
    "    plt.axvline(q, color = 'red') \n",
    "\n",
    "plt.show()\n",
    "\n",
    "trimmed_expression = gene_expression[(gene_expression>p_25)&(gene_expression<p_75)]\n",
    "trimmed_mean = np.mean(trimmed_expression)\n",
    "trimmed_mean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7215fd0a",
   "metadata": {},
   "source": [
    "Okay, so: \n",
    "\n",
    "- For each gene, find the expression distribution at the cluster level, find the cluster with the highest trimmed avg expression, have that for sst as the threshold, define for the rest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c972cf0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "X = reference.X\n",
    "if issparse(X) and not isinstance(X, csr_matrix):\n",
    "    X = X.tocsr()\n",
    "\n",
    "clusters = reference.obs['cluster_label'].to_numpy()\n",
    "uniq_clusters, inverse = np.unique(clusters, return_inverse=True)\n",
    "\n",
    "n_genes = reference.n_vars\n",
    "max_expression = np.full(n_genes, -np.inf, dtype=float)\n",
    "\n",
    "for k, cl in enumerate(uniq_clusters):\n",
    "    # rows for this cluster\n",
    "    rows = np.nonzero(inverse == k)[0]\n",
    "    if rows.size == 0:\n",
    "        continue\n",
    "\n",
    "    # submatrix (cells_in_cluster x genes), dense for vectorized percentile ops\n",
    "    Xc = X[rows, :].toarray() if issparse(X) else np.asarray(X[rows, :])\n",
    "\n",
    "    # per-gene quartiles (vectorized across genes)\n",
    "    p25, p75 = np.percentile(Xc, [25, 75], axis=0)\n",
    "\n",
    "    # trim per gene\n",
    "    mask = (Xc > p25) & (Xc < p75)            # shape: (cells_in_cluster, n_genes)\n",
    "    count_trim = mask.sum(axis=0)             # per-gene counts after trimming\n",
    "    sum_trim = (Xc * mask).sum(axis=0)        # per-gene sums after trimming\n",
    "\n",
    "    # avoid division-by-zero when all values fall outside (rare but possible)\n",
    "    trimmed_mean = np.divide(\n",
    "        sum_trim, count_trim,\n",
    "        out=np.zeros_like(sum_trim, dtype=float),\n",
    "        where=count_trim > 0\n",
    "    )\n",
    "\n",
    "    # update running max across clusters\n",
    "    np.maximum(max_expression, trimmed_mean, out=max_expression)\n",
    "\n",
    "reference.var['expression_max'] = max_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e133c8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(reference.var['expression_max'], bins = 200, log = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd266aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference.var[reference.var_names=='Sst']['expression_max']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b0126ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "sst_mask = reference.obs['subclass_label'] == 'Sst'\n",
    "sst_clusters = reference.obs['cluster_label'][sst_mask]\n",
    "\n",
    "print(sst_clusters.unique())\n",
    "\n",
    "expression = np.ones_like(sst_clusters.unique())\n",
    "ncells = np.ones_like(sst_clusters.unique())\n",
    "\n",
    "for i, cluster in enumerate(sst_clusters.unique()):\n",
    "\n",
    "    gene_idx = reference.var_names.get_loc('Sst')\n",
    "    print(gene_idx)\n",
    "    gene_expression = reference.X[:,gene_idx].toarray().flatten()\n",
    "\n",
    "    sst_mask = reference.obs['cluster_label'] == cluster\n",
    "    print(sum(sst_mask))\n",
    "    gene_expression = gene_expression[sst_mask]\n",
    "\n",
    "    #plt.hist(gene_expression, bins = 200, log = True)\n",
    "    [p_25, p_75] = [np.percentile(gene_expression, q) for q in [25, 75]]\n",
    "    print([p_25, p_75])\n",
    "    trimmed_expression = gene_expression[(gene_expression>p_25)&(gene_expression<p_75)]\n",
    "    trimmed_mean = np.mean(trimmed_expression)\n",
    "\n",
    "    expression[i] = trimmed_mean\n",
    "    ncells[i] = len(gene_expression)\n",
    "    #plt.title(f'Expression levels of Sst in {cluster} cells (counts) TM:{trimmed_mean}')\n",
    "\n",
    "    #for q in [p_25, p_75]:\n",
    "    #    plt.axvline(q, color = 'red') \n",
    "\n",
    "    #plt.show()\n",
    "\n",
    "plt.scatter(expression, ncells)\n",
    "plt.title('Expression of Sst per Sst cluster')\n",
    "plt.xlabel('Trimmed mean(counts)')\n",
    "plt.ylabel('ncells per cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7deb528a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to dense\n",
    "X = reference_filtered.X.toarray() if hasattr(reference_filtered.X, \"toarray\") else np.asarray(reference_filtered.X)\n",
    "\n",
    "# Flatten into 1D\n",
    "values = X.ravel()\n",
    "\n",
    "plt.hist(values, bins=200, log=True)\n",
    "plt.xlabel(\"Log-norm expression\")\n",
    "plt.ylabel(\"Frequency (log scale)\")\n",
    "plt.title(\"Histogram of all values in reference.X\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1528ae2f",
   "metadata": {},
   "source": [
    "Let's try to define a 1% expression level per gene, a la spapros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c96698",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_threshold = np.percentile(reference.X.toarray(), 99, axis = 0)\n",
    "plt.hist(max_threshold, bins = 200, log = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d085b35f",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference.var['max_threshold_spapros']= max_threshold\n",
    "reference.var[reference.var_names=='Sst']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e6fc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(reference.var.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cdc4f71",
   "metadata": {},
   "source": [
    "## Sparseness constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc2bb7a3",
   "metadata": {},
   "source": [
    "We want a gene to be expressed \"sparsely\". For that, we can either have an interesting metric, or we can have a boxplot. Let's start with the boxplot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "972bc94d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading the yao2021 reference\n",
    "\n",
    "ALM_subclasses = [\n",
    "    \"Vip\",\n",
    "    \"Lamp5\",\n",
    "    \"Scng\",\n",
    "    \"Sst Chodl\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"L2/3 IT CTX\",\n",
    "    \"L4/5 IT CTX\",\n",
    "    \"L5 IT CTX\",\n",
    "    \"L6 IT CTX\",\n",
    "    \"L5 PT CTX\",\n",
    "    \"L4 RSP-ACA\",\n",
    "    \"L5/6 NP CTX\",\n",
    "    \"L6 CT CTX\",\n",
    "    \"L6b CTX\",\n",
    "]\n",
    "\n",
    "\n",
    "datapath = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/yao_2021')\n",
    "\n",
    "reference = io.load_yao_2021_to_anndata(datapath, 'ALM', ALM_subclasses)\n",
    "counts_reference = reference.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0092928a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_genes(reference, min_cells=1) # it crashes with a lot of genes expressed in no cells\n",
    "sc.pp.normalize_total(reference)\n",
    "sc.pp.log1p(reference)\n",
    "sc.pp.highly_variable_genes(reference,flavor=\"cell_ranger\",n_top_genes=10000) #for a good panel\n",
    "\n",
    "sc.pp.filter_genes(counts_reference, min_cells=1) # it crashes with a lot of genes expressed in no cells\n",
    "sc.pp.highly_variable_genes(counts_reference,flavor=\"cell_ranger\",n_top_genes=10000) #for a good panel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22710285",
   "metadata": {},
   "outputs": [],
   "source": [
    "savepath = \"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/spapros_10K_taylored_glia_2021\"\n",
    "\n",
    "prior_marker = [\n",
    "    \"Chodl\",\n",
    "    \"Cux2\",\n",
    "    \"Fezf2\",\n",
    "    \"Foxp2\",\n",
    "    \"Rorb\",\n",
    "    \"Pvalb\",\n",
    "    \"Lamp5\",\n",
    "    \"Adamts2\",\n",
    "    \"Slco2a1\",\n",
    "]\n",
    "\n",
    "selector = sp.se.ProbesetSelector(reference, n=100, \n",
    "                                  celltype_key=\"subclass_label\", \n",
    "                                  verbosity=1, \n",
    "                                  n_jobs=-1, \n",
    "                                  save_dir=savepath, \n",
    "                                  preselected_genes=prior_marker)\n",
    "\n",
    "selected = selector.probeset[selector.probeset[\"selection\"]].copy()\n",
    "spapros_panel = list(selected.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b833b121",
   "metadata": {},
   "source": [
    "Subset the reference dataset by the genes we care about"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd5a425",
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_reference = reference[:,reference.var_names.isin(spapros_panel)]\n",
    "panel_counts_reference = counts_reference[:,reference.var_names.isin(spapros_panel)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e3b3c28",
   "metadata": {},
   "source": [
    "Calculate the trimmed mean per cluster of each gene. In raw counts or in CPM? Do it in their lognorm way. \n",
    "\n",
    "We want to have a matrix of clusters x genes, or add it to the reference.var. Let's add it to the var if it's less than 100 clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "028d12eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(panel_reference.obs['cluster_label'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81a0a377",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_expression_per_taxa(panel_reference, taxa_label):\n",
    "\n",
    "    '''\n",
    "    Generates a matrix of taxa x genes, where each entry corresponds to the trimmed (25-75) mean expression per taxa. \n",
    "    '''\n",
    "\n",
    "    #initialise the dataframe\n",
    "    genes = panel_reference.var_names\n",
    "\n",
    "    clusters = panel_reference.obs[taxa_label].to_numpy()\n",
    "    uniq_clusters, inverse = np.unique(clusters, return_inverse=True) #inverse shows which elements belong to which cluster in the original matrix\n",
    "\n",
    "    expression_percluster = pd.DataFrame(index=uniq_clusters, columns=genes)\n",
    "\n",
    "    n_genes = panel_reference.n_vars\n",
    "    print(f'There are {n_genes} genes in this panel')\n",
    "\n",
    "    for k, cl in enumerate(uniq_clusters):\n",
    "        # rows for this cluster\n",
    "        rows = np.nonzero(inverse == k)[0]\n",
    "        if rows.size == 0:\n",
    "            continue\n",
    "\n",
    "        # submatrix (cells_in_cluster x genes), dense for vectorized percentile ops\n",
    "        Xc = panel_reference.X[rows, :].toarray() if issparse(panel_reference.X) else np.asarray(panel_reference.X[rows, :])\n",
    "\n",
    "        # per-gene quartiles (vectorized across genes)\n",
    "        p25, p75 = np.percentile(Xc, [25, 75], axis=0)\n",
    "\n",
    "        # trim per gene\n",
    "        mask = (Xc > p25) & (Xc < p75)            # shape: (cells_in_cluster, n_genes)\n",
    "        count_trim = mask.sum(axis=0)             # per-gene counts after trimming\n",
    "        sum_trim = (Xc * mask).sum(axis=0)        # per-gene sums after trimming\n",
    "\n",
    "        # avoid division-by-zero when all values fall outside (rare but possible)\n",
    "        trimmed_mean = np.divide(\n",
    "            sum_trim, count_trim,\n",
    "            out=np.zeros_like(sum_trim, dtype=float),\n",
    "            where=count_trim > 0\n",
    "        )\n",
    "\n",
    "        # add to our matrix\n",
    "        expression_percluster.loc[cl] = trimmed_mean\n",
    "\n",
    "    return expression_percluster\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcd7e66a",
   "metadata": {},
   "source": [
    "This is how they're ordered in the panel, for easy comparison:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d659100c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is for the panel with no sparseness constraints\n",
    "\n",
    "gene_priority = [\n",
    "    \"Chodl\",\"Vip\",\"Rorb\",\"Pvalb\",\"Sst\",\"Fezf2\",\"Lamp5\",\"Cux2\",\"Rrad\",\"Adamts2\",\"Foxp2\",\"Npsr1\",\n",
    "    \"Slco2a1\",\"Olfm1\",\"Calb1\",\"Zfpm2\",\"Cdh13\",\"Cryab\",\"Etv1\",\"Cdh8\",\"Rgs4\",\"Grin3a\",\"Cplx3\",\"Gpc6\",\n",
    "    \"Synpr\",\"Kcnab1\",\"Cacna2d3\",\"Vwc2l\",\"Rprm\",\"Tox\",\"Erbb4\",\"Crh\",\"Ptn\",\"Lypd1\",\"Zfp804b\",\"Pcp4l1\",\n",
    "    \"Il1rapl2\",\"Lingo2\",\"Pcp4\",\"Nxph1\",\"Dcc\",\"Pcsk5\",\"Caln1\",\"Man1a\",\"Nov\",\"Prr16\",\"Grik1\",\"Cck\",\n",
    "    \"Cpne4\",\"S100a10\",\"Sorcs1\",\"Cdh9\",\"Gng4\",\"Pde1a\",\"Penk\",\"Alcam\",\"Reln\",\"Ptprt\",\"Trp53i11\",\"Marcksl1\",\n",
    "    \"Fstl5\",\"Igfbp4\",\"Cbln4\",\"Crhbp\",\"Necab1\",\"Crtac1\",\"Unc5d\",\"Col25a1\",\"Ptchd4\",\"Fxyd6\",\"Stxbp6\",\n",
    "    \"Thsd7a\",\"Parm1\",\"Sema3c\",\"Medag\",\"Rbp4\",\"Ctgf\",\"Chst8\",\"Fam19a1\",\"Adarb2\",\"Dner\",\"Ak5\",\"Cacng3\",\n",
    "    \"Cygb\",\"Nxph4\",\"Cplx1\",\"Arpp19\",\"Kcnk2\",\"Hs6st3\",\"Sparcl1\",\"Gpx3\",\"Car4\",\"Gucy1a3\",\"Spon1\",\"Prss23\",\n",
    "    \"Tac2\",\"Ptprm\",\"Tmem91\",\"Sema5a\",\"Cntnap5c\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "270afe56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is for box_sparseness\n",
    "\n",
    "gene_priority = [\n",
    "    \"Chodl\",\n",
    "    \"Vip\",\n",
    "    \"Rorb\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"Lamp5\",\n",
    "    \"Fezf2\",\n",
    "    \"Cux2\",\n",
    "    \"Rrad\",\n",
    "    \"Adamts2\",\n",
    "    \"Foxp2\",\n",
    "    \"Npsr1\",\n",
    "    \"Slco2a1\",\n",
    "    \"Tshz2\",\n",
    "    \"Fam19a1\",\n",
    "    \"Cplx3\",\n",
    "    \"Calb1\",\n",
    "    \"Hs3st4\",\n",
    "    \"Blnk\",\n",
    "    \"Igfbp6\",\n",
    "    \"Synpr\",\n",
    "    \"Adarb2\",\n",
    "    \"Ctgf\",\n",
    "    \"Syt6\",\n",
    "    \"Tnnc1\",\n",
    "    \"Erbb4\",\n",
    "    \"Sstr2\",\n",
    "    \"Cdh9\",\n",
    "    \"Ptprk\",\n",
    "    \"Stxbp6\",\n",
    "    \"Rprm\",\n",
    "    \"Slc30a3\",\n",
    "    \"Crh\",\n",
    "    \"Lypd1\",\n",
    "    \"Man1a\",\n",
    "    \"Rgs12\",\n",
    "    \"Npas1\",\n",
    "    \"Il1rapl2\",\n",
    "    \"Ptprm\",\n",
    "    \"Crym\",\n",
    "    \"Igfbp4\",\n",
    "    \"Kit\",\n",
    "    \"Cryab\",\n",
    "    \"Prss23\",\n",
    "    \"Etv1\",\n",
    "    \"Grin3a\",\n",
    "    \"Nxph1\",\n",
    "    \"Tcap\",\n",
    "    \"Meis2\",\n",
    "    \"Calb2\",\n",
    "    \"Pcp4l1\",\n",
    "    \"Egfr\",\n",
    "    \"Nr2f2\",\n",
    "    \"Grik1\",\n",
    "    \"Thsd7a\",\n",
    "    \"Cd63\",\n",
    "    \"Kcnip1\",\n",
    "    \"Tac2\",\n",
    "    \"Cbln4\",\n",
    "    \"Elfn1\",\n",
    "    \"Unc13c\",\n",
    "    \"Car4\",\n",
    "    \"Moxd1\",\n",
    "    \"Trp53i11\",\n",
    "    \"Cpne7\",\n",
    "    \"Rmst\",\n",
    "    \"Npy\",\n",
    "    \"Cartpt\",\n",
    "    \"Ndnf\",\n",
    "    \"Pthlh\",\n",
    "    \"S100b\",\n",
    "    \"Vwc2l\",\n",
    "    \"Crhbp\",\n",
    "    \"Arhgap25\",\n",
    "    \"Cox6a2\",\n",
    "    \"Gad1\",\n",
    "    \"Adamts3\",\n",
    "    \"Prex1\",\n",
    "    \"Npr3\",\n",
    "    \"Adcyap1\",\n",
    "    \"Sema3c\",\n",
    "    \"Sema5a\",\n",
    "    \"Pamr1\",\n",
    "    \"Npy2r\",\n",
    "    \"Penk\",\n",
    "    \"Fam46a\",\n",
    "    \"Kctd8\",\n",
    "    \"Pnoc\",\n",
    "    \"Cidea\",\n",
    "    \"Ackr3\",\n",
    "    \"Lgals1\",\n",
    "    \"Ptprz1\",\n",
    "    \"Cxcl14\",\n",
    "    \"Tcerg1l\",\n",
    "    \"Lypd6\",\n",
    "    \"Rxfp1\",\n",
    "    \"Serpine2\",\n",
    "    \"Reln\",\n",
    "    \"Cdh20\",\n",
    "    \"Timp3\",\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0006f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this is for the entropy constraint\n",
    "gene_priority = [\n",
    "    \"Chodl\",\n",
    "    \"Vip\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"Rorb\",\n",
    "    \"Lamp5\",\n",
    "    \"Cux2\",\n",
    "    \"Fezf2\",\n",
    "    \"Rrad\",\n",
    "    \"Adamts2\",\n",
    "    \"Foxp2\",\n",
    "    \"Npsr1\",\n",
    "    \"Slco2a1\",\n",
    "    \"Tshz2\",\n",
    "    \"Fam19a1\",\n",
    "    \"Calb1\",\n",
    "    \"Syt6\",\n",
    "    \"Blnk\",\n",
    "    \"Cplx3\",\n",
    "    \"Adarb2\",\n",
    "    \"Tcap\",\n",
    "    \"Rprm\",\n",
    "    \"Cpne9\",\n",
    "    \"Tac1\",\n",
    "    \"Ctgf\",\n",
    "    \"Erbb4\",\n",
    "    \"Slc30a3\",\n",
    "    \"Synpr\",\n",
    "    \"Cdh9\",\n",
    "    \"Tnnc1\",\n",
    "    \"Il1rapl2\",\n",
    "    \"Meis2\",\n",
    "    \"Crh\",\n",
    "    \"Adcyap1\",\n",
    "    \"Rgs12\",\n",
    "    \"Lypd1\",\n",
    "    \"Igfbp6\",\n",
    "    \"Sema5a\",\n",
    "    \"Cpne7\",\n",
    "    \"S100a6\",\n",
    "    \"Lypd6\",\n",
    "    \"Calb2\",\n",
    "    \"Pamr1\",\n",
    "    \"Cidea\",\n",
    "    \"Medag\",\n",
    "    \"Ptprk\",\n",
    "    \"Pcp4l1\",\n",
    "    \"Nxph1\",\n",
    "    \"Crym\",\n",
    "    \"Ptprm\",\n",
    "    \"Cbln4\",\n",
    "    \"Igfbp4\",\n",
    "    \"Pnoc\",\n",
    "    \"Penk\",\n",
    "    \"Npy\",\n",
    "    \"Egfr\",\n",
    "    \"Cryab\",\n",
    "    \"Grik1\",\n",
    "    \"Kctd8\",\n",
    "    \"Car4\",\n",
    "    \"S100b\",\n",
    "    \"Nr2f2\",\n",
    "    \"Pthlh\",\n",
    "    \"Prss23\",\n",
    "    \"Cdh20\",\n",
    "    \"Ptprz1\",\n",
    "    \"Vwc2l\",\n",
    "    \"Elfn1\",\n",
    "    \"Crhbp\",\n",
    "    \"Gad1\",\n",
    "    \"Angpt1\",\n",
    "    \"Adamts3\",\n",
    "    \"Nxph4\",\n",
    "    \"Npr3\",\n",
    "    \"Sema5b\",\n",
    "    \"Ddit4l\",\n",
    "    \"Kit\",\n",
    "    \"Serpine2\",\n",
    "    \"Unc13c\",\n",
    "    \"Cd63\",\n",
    "    \"Cartpt\",\n",
    "    \"Thsd7a\",\n",
    "    \"Tac2\",\n",
    "    \"Moxd1\",\n",
    "    \"Rmst\",\n",
    "    \"Sema3c\",\n",
    "    \"Rspo1\",\n",
    "    \"Sox6\",\n",
    "    \"Ppapdc1a\",\n",
    "    \"Lgals1\",\n",
    "    \"Ackr3\",\n",
    "    \"Timp3\",\n",
    "    \"Npas1\",\n",
    "    \"Reln\",\n",
    "    \"Vwc2\",\n",
    "    \"Pcdh8\",\n",
    "    \"Hgf\",\n",
    "    \"Pcdh20\",\n",
    "    \"Cp\",\n",
    "    \"Ndnf\",\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ca2ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_percluster = ut.calculate_expression_per_taxa(panel_reference, 'cluster_label')\n",
    "\n",
    "# --- Build robust cluster → subclass mapping from obs ---\n",
    "meta = (\n",
    "    reference.obs[[\"cluster_label\", \"subclass_label\"]]\n",
    "    .dropna()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "ALM_subclasses = reference.obs['subclass_label'].unique()\n",
    "\n",
    "# ---- Explicit color allocation (edit as you wish) ----\n",
    "subclass_colors = {\n",
    "    \"Vip\":           \"#1f77b4\",\n",
    "    \"Lamp5\":         \"#ff7f0e\",\n",
    "    \"Scng\":          \"#2ca02c\",\n",
    "    \"Sst Chodl\":     \"#d62728\",\n",
    "    \"Sst\":           \"#9467bd\",\n",
    "    \"Pvalb\":         \"#8c564b\",\n",
    "    \"L2/3 IT CTX\":   \"#e377c2\",\n",
    "    \"L4/5 IT CTX\":   \"#7f7f7f\",\n",
    "    \"L5 IT CTX\":     \"#bcbd22\",\n",
    "    \"L6 IT CTX\":     \"#17becf\",\n",
    "    \"L5 PT CTX\":     \"#aec7e8\",\n",
    "    \"L4 RSP-ACA\":    \"#ffbb78\",\n",
    "    \"L5/6 NP CTX\":   \"#98df8a\",\n",
    "    \"L6 CT CTX\":     \"#ff9896\",\n",
    "    \"L6b CTX\":       \"#c5b0d5\",\n",
    "}\n",
    "\n",
    "# Default color for any subclass not listed\n",
    "default_color = \"#dddddd\"\n",
    "\n",
    "# ---- Order clusters by subclass group, then numeric id within group ----\n",
    "ordered_clusters = []\n",
    "spans = []  # list of (subclass, start_idx, end_idx) for shading bands\n",
    "\n",
    "start = 0\n",
    "for sub in ALM_subclasses:\n",
    "    members = meta.loc[meta[\"subclass_label\"] == sub].sort_values(\"cluster_label\", kind=\"mergesort\")\n",
    "    if not members.empty:\n",
    "        ordered_clusters.extend(members[\"cluster_label\"].tolist())\n",
    "        end = start + len(members) - 1\n",
    "        spans.append((sub, start, end))\n",
    "        start = end + 1\n",
    "\n",
    "# Reindex the dataframe to the new cluster and gene order\n",
    "# 2) Compute the new column order:\n",
    "cols = list(expression_percluster.columns)\n",
    "prio_set = set(gene_priority)\n",
    "\n",
    "# first: the genes you specified, in exactly that order (and present in df)\n",
    "ordered_cols = [g for g in gene_priority if g in cols]\n",
    "# then: any remaining df columns not in your list, preserving their original order\n",
    "ordered_cols += [g for g in cols if g not in prio_set]\n",
    "\n",
    "# 3) Reindex the dataframe columns accordingly (this returns a new df)\n",
    "expression_percluster_ordered = expression_percluster.loc[:, ordered_cols]\n",
    "\n",
    "#and now, per clusters\n",
    "expr_ord = expression_percluster_ordered.loc[ordered_clusters]\n",
    "\n",
    "# ---- Plot: one subplot (row) per gene; shaded subclass bands ----\n",
    "genes = list(expr_ord.columns)\n",
    "expr_ord.max()\n",
    "n_genes = len(genes)\n",
    "fig_h = max(2 * n_genes, 4)\n",
    "\n",
    "# ---- Layout knobs ----\n",
    "height_per_subplot = 4.4   # bump this up if still too squished\n",
    "fig_w = 12\n",
    "legend_right_margin = 0.80 # reserve space for legend (0.80 = 80% of width for plots)\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "    n_genes, 1,\n",
    "    figsize=(fig_w, max(4, height_per_subplot * n_genes)),\n",
    "    sharex=True\n",
    ")\n",
    "\n",
    "\n",
    "if n_genes == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "x = np.arange(len(ordered_clusters))\n",
    "\n",
    "for i, gene in enumerate(genes):\n",
    "    ax = axes[i]\n",
    "\n",
    "    # background subclass bands\n",
    "    for sub, s, e in spans:\n",
    "        color = subclass_colors.get(sub, default_color)\n",
    "        ax.axvspan(s - 0.5, e + 0.5, facecolor=color, alpha=0.12, linewidth=0)\n",
    "\n",
    "    # gene trace\n",
    "    y = expr_ord[gene].astype(float).to_numpy()\n",
    "    ax.plot(x, y, marker=\"o\", linestyle=\"-\", linewidth=1)\n",
    "    ax.set_ylabel(f'log1p(norm({gene}))')\n",
    "    ax.grid(True, alpha=0.3)\n",
    "\n",
    "# X ticks\n",
    "axes[-1].set_xticks(x)\n",
    "axes[-1].set_xticklabels(ordered_clusters, rotation=90)\n",
    "axes[-1].set_xlabel(\"Clusters (grouped by subclass)\")\n",
    "\n",
    "# Legend: one entry per subclass band that actually appears\n",
    "legend_patches = []\n",
    "for sub, _, _ in spans:\n",
    "    if sub not in [p.get_label() for p in legend_patches]:\n",
    "        legend_patches.append(Patch(facecolor=subclass_colors.get(sub, default_color), alpha=0.4, label=sub))\n",
    "# Put legend above or to the side to avoid overlapping plots\n",
    "for ax in axes:\n",
    "    ax.legend(handles=legend_patches, loc=\"upper left\", bbox_to_anchor=(1.01, 1.0), frameon=False, title=\"Subclass bands\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# If you also want to keep the explicit order for later use:\n",
    "ordered_clusters_list = ordered_clusters\n",
    "subclass_band_spans = spans  # [(subclass, start_idx, end_idx), ...]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d5d91e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_percluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b3ce90",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = 100  # control rows\n",
    "pd.options.display.max_columns = 10  # control cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9627249a",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_persubclass = ut.calculate_expression_per_taxa(panel_reference, 'subclass_label')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4efce8e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(list(expression_persubclass.columns)==sorted(list(expression_persubclass.columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64c40e5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(expression_persubclass.values[:,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f2f9ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_persubclass = ut.calculate_expression_per_taxa(panel_reference, 'subclass_label')\n",
    "\n",
    "\n",
    "maxmean = expression_persubclass.max() - expression_persubclass.mean()\n",
    "\n",
    "#calculate entropy of each gene\n",
    "gene_H = entropy(expression_persubclass.to_numpy(dtype = float), base = 2, axis = 0)\n",
    "\n",
    "h_thresh = 3.0        # entropy threshold (bits)\n",
    "mm_thresh = 2.0       # max-mean threshold (same units as your data)\n",
    "mask = (gene_H > h_thresh) & (maxmean < mm_thresh)\n",
    "almost_mask = (gene_H > h_thresh) ^ (maxmean < mm_thresh)\n",
    "colors_1 = np.where(mask, 1, 0)\n",
    "colors_2 = np.where(almost_mask, 2, 0)\n",
    "colors_sum = colors_1+colors_2\n",
    "colist = ['black', 'red', 'orange']\n",
    "colors = [colist[i] for i in colors_sum]\n",
    "\n",
    "plt.scatter(maxmean, gene_H, c=colors, s=20)\n",
    "plt.ylabel('per gene Shannon entropy (bits)')\n",
    "plt.xlabel('per gene max-mean (log1p(norm(counts))')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fff75c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def logistic(x, k, x0, plot = False):\n",
    "    y = [1/(1+np.power(np.e, -k*(value-x0))) for value in x]\n",
    "    if plot:\n",
    "        plt.scatter(x, y)\n",
    "        plt.title('Logistic for midpoint {x0} and k {k}')\n",
    "        plt.show()\n",
    "    return y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a7a44d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.cm import get_cmap\n",
    "\n",
    "values = np.linspace(1, 5, 100)\n",
    "print(values)\n",
    "ks = np.linspace(0, 5, 10)\n",
    "print(ks)\n",
    "cmap = get_cmap(\"viridis\", 10)           # 10 discrete steps of 'viridis'\n",
    "\n",
    "for i, k in enumerate(ks):\n",
    "    y = np.array(logistic(values, k, 2))\n",
    "    plt.scatter(values, 1-y, color = cmap(i))\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61b054c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#assert that gene_priority is just a resorting of genes\n",
    "\n",
    "for gene in genes:\n",
    "    assert gene in gene_priority, f'{gene}'\n",
    "assert len(genes)==len(gene_priority)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "193d5410",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# --- Build robust cluster → subclass mapping from obs ---\n",
    "\n",
    "\n",
    "# ---- Explicit color allocation (edit as you wish) ----\n",
    "subclass_colors = {\n",
    "    \"Vip\":           \"#1f77b4\",\n",
    "    \"Lamp5\":         \"#ff7f0e\",\n",
    "    \"Scng\":          \"#2ca02c\",\n",
    "    \"Sst Chodl\":     \"#d62728\",\n",
    "    \"Sst\":           \"#9467bd\",\n",
    "    \"Pvalb\":         \"#8c564b\",\n",
    "    \"L2/3 IT CTX\":   \"#e377c2\",\n",
    "    \"L4/5 IT CTX\":   \"#7f7f7f\",\n",
    "    \"L5 IT CTX\":     \"#bcbd22\",\n",
    "    \"L6 IT CTX\":     \"#17becf\",\n",
    "    \"L5 PT CTX\":     \"#aec7e8\",\n",
    "    \"L4 RSP-ACA\":    \"#ffbb78\",\n",
    "    \"L5/6 NP CTX\":   \"#98df8a\",\n",
    "    \"L6 CT CTX\":     \"#ff9896\",\n",
    "    \"L6b CTX\":       \"#c5b0d5\",\n",
    "}\n",
    "\n",
    "expr_ord = expression_persubclass\n",
    "# ---- Plot: one subplot (row) per gene; shaded subclass bands ----\n",
    "genes_order =np.argsort((expr_ord.max() - expr_ord.mean()).values)\n",
    "\n",
    "genes = list(expr_ord.columns[genes_order]) #sort genes according to max-mean quality metric\n",
    "\n",
    "plot_maxmean = maxmean[np.argsort(maxmean)]#sort metrics according to the same metric to preserve correspondence\n",
    "plot_entropy = gene_H[np.argsort(maxmean)]\n",
    "plot_colors = np.array(colors)[np.argsort(maxmean)]\n",
    "\n",
    "n_genes = len(genes)\n",
    "\n",
    "# ---- Layout knobs ----\n",
    "height_per_subplot = 4.4   # bump this up if still too squished\n",
    "fig_w = 12\n",
    "legend_right_margin = 0.80 # reserve space for legend (0.80 = 80% of width for plots)\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "    n_genes, 1,\n",
    "    figsize=(fig_w, max(4, height_per_subplot * n_genes)),\n",
    ")\n",
    "\n",
    "\n",
    "if n_genes == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "x = np.arange(len(reference.obs['subclass_label'].unique()))\n",
    "\n",
    "for i, gene in enumerate(genes):\n",
    "    ax = axes[i]\n",
    "    # gene trace\n",
    "    y = expr_ord[gene].astype(float).to_numpy()\n",
    "    ax.plot(x, y, marker=\"o\", linestyle=\"none\", markersize=20, linewidth=1)\n",
    "    ax.text(0.99, 0.95,\n",
    "        f'Entropy: {plot_entropy[i]:.3f}\\nMax-mean: {plot_maxmean[i]:.3f}\\n{gene}',\n",
    "        transform=ax.transAxes,  # <--- relative to subplot axes\n",
    "        ha=\"right\", va=\"top\", \n",
    "        color = plot_colors[i])\n",
    "\n",
    "    ax.set_ylabel(f'log1p(norm({gene}))')\n",
    "    ax.set_xticks(np.arange(len(expr_ord.index)), labels=expr_ord.index)\n",
    "    ax.grid(True, alpha=0.3)\n",
    "    ax.set_ylim(-0.5, 10)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe76129",
   "metadata": {},
   "source": [
    "We're going to train on hand-picked genes to see how well we capture it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8873e5c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "train = pd.read_csv('sparseness_train.csv')\n",
    "result = [0, 1, 0] #translate 1, 2 into binary mask\n",
    "train['target'] = [result[i] for i in train['Sparseness']]\n",
    "train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c1f2890",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort to match the metric. Metrics are in reference order, which is alphabetical\n",
    "train = train.sort_values(by='Gene')\n",
    "assert list(train['Gene'])==list(expression_persubclass.columns), 'sorting went wrong somewhere'\n",
    "train['entropy'] = gene_H\n",
    "train['maxmean'] = maxmean.values\n",
    "train = train.dropna(subset=[\"entropy\", \"maxmean\"])\n",
    "train.head\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3848507",
   "metadata": {},
   "outputs": [],
   "source": [
    "color = np.where(train['target']==0, \"red\", \"green\")\n",
    "plt.scatter(train['maxmean'], train['entropy'], color = color, s = 20)\n",
    "plt.ylabel('per gene Shannon entropy (bits)')\n",
    "plt.xlabel('per gene max-mean (log1p(norm(counts))')\n",
    "\n",
    "plt.axhline(3.28, color = 'lightblue')\n",
    "plt.axvline(3.38, color = 'lightgray')\n",
    "plt.axvline(2, color = 'lightblue')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00cc61df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, classification_report\n",
    "\n",
    "# Features and target\n",
    "X = train[[\"entropy\", \"maxmean\"]]\n",
    "y = train[\"target\"]\n",
    "\n",
    "# Split into train/test sets (80/20 for example)\n",
    "#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Initialize logistic regression\n",
    "full_model = LogisticRegression(max_iter=1000)\n",
    "\n",
    "# Fit model\n",
    "full_model.fit(X, y)\n",
    "\n",
    "# Predict on test set\n",
    "#y_pred = clf.predict(X_test)\n",
    "\n",
    "# Evaluate\n",
    "#print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
    "#print(classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2a2a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Intercept:\", full_model.intercept_)\n",
    "print(\"Coefficients:\", dict(zip(X.columns, full_model.coef_[0])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c16d4ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.DataFrame(np.linspace(0, 7, 500))\n",
    "score = full_model.predict_proba(test)\n",
    "print(score.shape)\n",
    "plt.plot(test, score[:,1])\n",
    "plt.ylabel('Gene score')\n",
    "plt.xlabel('Gene entropy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09437646",
   "metadata": {},
   "outputs": [],
   "source": [
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce570bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import cross_val_score, StratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "import numpy as np\n",
    "\n",
    "def cv_score(features, k=5):\n",
    "    X_sub = train[features]\n",
    "    y = train[\"target\"]\n",
    "\n",
    "    clf = LogisticRegression(max_iter=1000)\n",
    "\n",
    "    # StratifiedKFold preserves class balance in splits\n",
    "    cv = StratifiedKFold(n_splits=k, shuffle=True, random_state=42)\n",
    "\n",
    "    scores = cross_val_score(clf, X_sub, y, cv=cv, scoring=\"roc_auc\")\n",
    "    return scores.mean(), scores.std()\n",
    "\n",
    "print(\"Both:\", cv_score([\"entropy\", \"maxmean\"]))\n",
    "print(\"Entropy only:\", cv_score([\"entropy\"]))\n",
    "print(\"Maxmean only:\", cv_score([\"maxmean\"]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a115d801",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.tree import plot_tree\n",
    "\n",
    "\n",
    "# Drop NaNs in predictors + target\n",
    "df = train.dropna(subset=[\"entropy\", \"maxmean\", \"target\"])\n",
    "\n",
    "X = df[[\"entropy\", \"maxmean\"]]\n",
    "y = df[\"target\"]\n",
    "\n",
    "# Split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train decision tree\n",
    "tree_clf = DecisionTreeClassifier(max_depth=2, random_state=42)  # tune max_depth if needed\n",
    "tree_clf.fit(X_train, y_train)\n",
    "\n",
    "# Evaluate\n",
    "y_pred = tree_clf.predict(X_test)\n",
    "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
    "\n",
    "# Feature importances\n",
    "importances = tree_clf.feature_importances_\n",
    "for feat, imp in zip(X.columns, importances):\n",
    "    print(f\"{feat}: {imp:.3f}\")\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12, 6))\n",
    "plot_tree(\n",
    "    tree_clf,\n",
    "    feature_names=X.columns,\n",
    "    class_names=[str(c) for c in tree_clf.classes_],\n",
    "    filled=True,\n",
    "    rounded=True,\n",
    "    fontsize=10\n",
    ")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06fc3894",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "\n",
    "df = train.dropna(subset=[\"entropy\", \"maxmean\", \"target\"])\n",
    "X = df[[\"maxmean\"]]\n",
    "y = df[\"target\"]\n",
    "\n",
    "# Define the parameter grid\n",
    "param_grid = {\n",
    "    \"max_depth\": [1],\n",
    "    \"min_samples_split\": [2, 5, 10],\n",
    "    \"min_samples_leaf\": [1, 2, 4]\n",
    "}\n",
    "\n",
    "clf = DecisionTreeClassifier(random_state=42)\n",
    "\n",
    "grid_search = GridSearchCV(\n",
    "    clf,\n",
    "    param_grid,\n",
    "    cv=5,               # 5-fold cross-validation\n",
    "    scoring=\"roc_auc\", # or \"roc_auc\"\n",
    "    n_jobs=-1\n",
    ")\n",
    "\n",
    "grid_search.fit(X, y)\n",
    "\n",
    "print(\"Best params:\", grid_search.best_params_)\n",
    "print(\"Best CV score:\", grid_search.best_score_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6b42a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from sklearn.tree import plot_tree\n",
    "\n",
    "best_tree = grid_search.best_estimator_\n",
    "\n",
    "plt.figure(figsize=(12, 6))\n",
    "plot_tree(\n",
    "    best_tree,\n",
    "    feature_names=X.columns,\n",
    "    class_names=[str(c) for c in best_tree.classes_],\n",
    "    filled=True,\n",
    "    rounded=True,\n",
    "    fontsize=10\n",
    ")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5f3478",
   "metadata": {},
   "source": [
    "### Applying final constraint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0c0ae2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "expression_persubclass = ut.calculate_expression_per_taxa(panel_reference, 'subclass_label')\n",
    "\n",
    "## Calculating metrics\n",
    "maxmean = expression_persubclass.max() - expression_persubclass.mean()\n",
    "#calculate entropy of each gene\n",
    "gene_H = entropy(expression_persubclass.to_numpy(dtype = float), base = 2, axis = 0)\n",
    "\n",
    "## Building train dataset\n",
    "train = pd.read_csv('sparseness_train.csv')\n",
    "result = [0, 1, 0] #translate 1, 2 into binary mask\n",
    "train['target'] = [result[i] for i in train['Sparseness']]\n",
    "\n",
    "#sort to match the metric. Metrics are in reference order, which is alphabetical\n",
    "train = train.sort_values(by='Gene')\n",
    "assert list(train['Gene'])==list(expression_persubclass.columns), 'sorting went wrong somewhere'\n",
    "train['entropy'] = gene_H\n",
    "train['maxmean'] = maxmean.values\n",
    "train = train.dropna(subset=[\"entropy\", \"maxmean\"])\n",
    "\n",
    "##Training final model\n",
    "# Features and target\n",
    "X = train[[\"entropy\"]]\n",
    "y = train[\"target\"]\n",
    "# Initialize logistic regression\n",
    "full_model = LogisticRegression(max_iter=1000)\n",
    "# Fit model\n",
    "full_model.fit(X, y)\n",
    "print(\"Intercept:\", full_model.intercept_)\n",
    "print(\"Coefficients:\", dict(zip(X.columns, full_model.coef_[0])))\n",
    "\n",
    "#Evaluate all genes\n",
    "all_genes_subclass = ut.calculate_expression_per_taxa(reference, 'subclass_label')\n",
    "all_genes_H = entropy(all_genes_subclass.to_numpy(dtype = float), base = 2, axis = 0)\n",
    "assert list(reference.var_names)==list(all_genes_subclass.columns), 'Something is wrong with the sorting'\n",
    "all_genes_entropies = pd.DataFrame(all_genes_H, index=reference.var_names)\n",
    "all_genes_entropies = all_genes_entropies.fillna(100) \n",
    "score = full_model.predict_proba(all_genes_entropies)\n",
    "reference.var['score'] = score[:,0]\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7730a6c6",
   "metadata": {},
   "source": [
    "Save parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67afdac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"intercept\": full_model.intercept_.tolist(),\n",
    "    \"coef\": full_model.coef_.tolist(),\n",
    "    \"features\": X.columns.tolist()\n",
    "}\n",
    "pd.Series(params).to_json(\"logreg_params.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04018ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_maxmean = all_genes_subclass.max() - all_genes_subclass.mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05be1c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def box_penalty(entropy, maxmean, k=5, entropy_threshold = 3.28, maxmean_threshold = 2):\n",
    "    \n",
    "    z_entropy = 1-np.array([1/(1+np.power(np.e, -k*(value-entropy_threshold))) for value in entropy])\n",
    "    z_maxmean= [1/(1+np.power(np.e, -k*(value-maxmean_threshold))) for value in maxmean]\n",
    "    penalty = np.minimum(z_entropy, z_maxmean)\n",
    "    return penalty\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9902571b",
   "metadata": {},
   "outputs": [],
   "source": [
    "penalty = box_penalty(all_genes_H, all_genes_maxmean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd3394c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_entropies = all_genes_entropies.rename(columns={0:'entropy'})\n",
    "all_genes_entropies['maxmean'] = all_genes_maxmean\n",
    "X = all_genes_entropies[[\"entropy\", 'maxmean']]\n",
    "score = full_model.predict_proba(X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75bd6bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "score = full_model.predict_proba(all_genes_entropies)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5bcf22a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Assume you already have:\n",
    "# all_genes_maxmean   -> array-like of shape (n_genes,)\n",
    "# all_genes_H         -> array-like of shape (n_genes,)\n",
    "# score               -> output of predict_proba (n_genes x 2)\n",
    "\n",
    "#probs = score[:, 1]  # probability of class=1\n",
    "probs = penalty\n",
    "\n",
    "plt.figure(figsize=(7, 6))\n",
    "sc = plt.scatter(\n",
    "    all_genes_maxmean,\n",
    "    all_genes_H,\n",
    "    c=probs,\n",
    "    cmap='Spectral',      # red=low, blue=high\n",
    "    s=20,\n",
    "    edgecolor=\"none\"\n",
    ")\n",
    "\n",
    "for e in entropies:\n",
    "    plt.axhline(e, color = \"lightgray\")\n",
    "\n",
    "plt.colorbar(sc, label=\"Penalty\")\n",
    "plt.xlabel(\"per gene max-mean (log1p(norm(counts)))\")\n",
    "plt.ylabel(\"per gene Shannon entropy (bits)\")\n",
    "plt.title(\"Gene distribution colored by box penalty score\")\n",
    "plt.grid(alpha=0.3)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df12c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(all_genes_H, log = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14ad9cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mask genes\n",
    "mask = (all_genes_H < 1) & (all_genes_maxmean < 1) & (all_genes_H>0.1)\n",
    "interesting_subset = all_genes_subclass.loc[:, mask]\n",
    "print(\"Subset shape:\", interesting_subset.shape)\n",
    "\n",
    "# Reproducible random sample\n",
    "np.random.seed(42)\n",
    "sampled_genes = np.random.choice(\n",
    "    interesting_subset.columns,\n",
    "    size=min(50, interesting_subset.shape[1]),\n",
    "    replace=False\n",
    ")\n",
    "\n",
    "# Subset the expression matrix\n",
    "sampled_subset = interesting_subset.loc[:, sampled_genes]\n",
    "\n",
    "# Subset entropy and maxmean consistently\n",
    "sampled_H = pd.Series(all_genes_H, index=all_genes_subclass.columns).loc[sampled_genes]\n",
    "sampled_maxmean = pd.Series(all_genes_maxmean, index=all_genes_subclass.columns).loc[sampled_genes]\n",
    "\n",
    "print(\"Sampled genes:\", list(sampled_genes)[:5], \"...\")\n",
    "print(\"Entropy subset shape:\", sampled_H.shape)\n",
    "print(\"Maxmean subset shape:\", sampled_maxmean.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7903b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "expr_ord = sampled_subset\n",
    "# ---- Plot: one subplot (row) per gene; shaded subclass bands ----\n",
    "genes_order =np.argsort((expr_ord.max() - expr_ord.mean()).values)\n",
    "\n",
    "genes = list(expr_ord.columns[genes_order]) #sort genes according to max-mean quality metric\n",
    "\n",
    "plot_maxmean = sampled_maxmean[np.argsort(sampled_maxmean)]#sort metrics according to the same metric to preserve correspondence\n",
    "plot_entropy = sampled_H[np.argsort(sampled_maxmean)]\n",
    "#plot_colors = np.array(colors)[np.argsort(maxmean)]\n",
    "\n",
    "n_genes = len(genes)\n",
    "\n",
    "# ---- Layout knobs ----\n",
    "height_per_subplot = 4.4   # bump this up if still too squished\n",
    "fig_w = 12\n",
    "legend_right_margin = 0.80 # reserve space for legend (0.80 = 80% of width for plots)\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "    n_genes, 1,\n",
    "    figsize=(fig_w, max(4, height_per_subplot * n_genes)),\n",
    ")\n",
    "\n",
    "\n",
    "if n_genes == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "x = np.arange(len(reference.obs['subclass_label'].unique()))\n",
    "\n",
    "for i, gene in enumerate(genes):\n",
    "    ax = axes[i]\n",
    "    # gene trace\n",
    "    y = expr_ord[gene].astype(float).to_numpy()\n",
    "    ax.plot(x, y, marker=\"o\", linestyle=\"none\", markersize=20, linewidth=1)\n",
    "    ax.text(0.99, 0.95,\n",
    "        f'Entropy: {plot_entropy[i]:.3f}\\nMax-mean: {plot_maxmean[i]:.3f}\\n{gene}',\n",
    "        transform=ax.transAxes,  # <--- relative to subplot axes\n",
    "        ha=\"right\", va=\"top\")\n",
    "\n",
    "    ax.set_ylabel(f'log1p(norm({gene}))')\n",
    "    ax.set_xticks(np.arange(len(expr_ord.index)), labels=expr_ord.index)\n",
    "    ax.grid(True, alpha=0.3)\n",
    "    ax.set_ylim(-0.5, 10)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63d7cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_subclasses = np.arange(1, 14)  # 1 to 13 inclusive\n",
    "\n",
    "sequences = []\n",
    "for n in n_subclasses:\n",
    "    seq = np.zeros(13)\n",
    "    seq[:n] = 1.0 / n     # put n values of 1/n at the start\n",
    "    sequences.append(seq)\n",
    "\n",
    "# If you want it as a NumPy array (13x13 matrix):\n",
    "sequences_array = np.vstack(sequences)\n",
    "\n",
    "entropies = np.zeros(13)\n",
    "\n",
    "for i, sequence in enumerate(sequences_array):\n",
    "    H1 = entropy(sequence, base = 2)\n",
    "    entropies[i] = H1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6282d595",
   "metadata": {},
   "outputs": [],
   "source": [
    "entropies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ede0b5e4",
   "metadata": {},
   "source": [
    "## Plot cummulative expression\n",
    "\n",
    "Find the average total read count per cluster. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a528d9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_ord.iloc[1,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49abb9b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "expression_percluster = ut.calculate_expression_per_taxa(panel_reference, 'cluster_label')\n",
    "\n",
    "# --- Build robust cluster → subclass mapping from obs ---\n",
    "meta = (\n",
    "    reference.obs[[\"cluster_label\", \"subclass_label\"]]\n",
    "    .dropna()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "ALM_subclasses = reference.obs['subclass_label'].unique()\n",
    "\n",
    "# ---- Explicit color allocation (edit as you wish) ----\n",
    "subclass_colors = {\n",
    "    \"Vip\":           \"#1f77b4\",\n",
    "    \"Lamp5\":         \"#ff7f0e\",\n",
    "    \"Scng\":          \"#2ca02c\",\n",
    "    \"Sst Chodl\":     \"#d62728\",\n",
    "    \"Sst\":           \"#9467bd\",\n",
    "    \"Pvalb\":         \"#8c564b\",\n",
    "    \"L2/3 IT CTX\":   \"#e377c2\",\n",
    "    \"L4/5 IT CTX\":   \"#7f7f7f\",\n",
    "    \"L5 IT CTX\":     \"#bcbd22\",\n",
    "    \"L6 IT CTX\":     \"#17becf\",\n",
    "    \"L5 PT CTX\":     \"#aec7e8\",\n",
    "    \"L4 RSP-ACA\":    \"#ffbb78\",\n",
    "    \"L5/6 NP CTX\":   \"#98df8a\",\n",
    "    \"L6 CT CTX\":     \"#ff9896\",\n",
    "    \"L6b CTX\":       \"#c5b0d5\",\n",
    "}\n",
    "\n",
    "# Default color for any subclass not listed\n",
    "default_color = \"#dddddd\"\n",
    "\n",
    "# ---- Order clusters by subclass group, then numeric id within group ----\n",
    "ordered_clusters = []\n",
    "spans = []  # list of (subclass, start_idx, end_idx) for shading bands\n",
    "\n",
    "start = 0\n",
    "for sub in ALM_subclasses:\n",
    "    members = meta.loc[meta[\"subclass_label\"] == sub].sort_values(\"cluster_label\", kind=\"mergesort\")\n",
    "    if not members.empty:\n",
    "        ordered_clusters.extend(members[\"cluster_label\"].tolist())\n",
    "        end = start + len(members) - 1\n",
    "        spans.append((sub, start, end))\n",
    "        start = end + 1\n",
    "\n",
    "# Reindex the dataframe to the new cluster order\n",
    "expr_ord = expression_percluster.loc[ordered_clusters]\n",
    "\n",
    "\n",
    "# ---- Layout knobs ----\n",
    "height_per_subplot = 4.4   # bump this up if still too squished\n",
    "fig_w = 12\n",
    "legend_right_margin = 0.80 # reserve space for legend (0.80 = 80% of width for plots)\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "    1, 1,\n",
    "    figsize=(fig_w, max(4, height_per_subplot ))\n",
    ")\n",
    "\n",
    "\n",
    "x = np.arange(len(ordered_clusters))\n",
    "\n",
    "\n",
    "ax = axes\n",
    "\n",
    "# background subclass bands\n",
    "for sub, s, e in spans:\n",
    "    color = subclass_colors.get(sub, default_color)\n",
    "    ax.axvspan(s - 0.5, e + 0.5, facecolor=color, alpha=0.12, linewidth=0)\n",
    "\n",
    "# expression_trace \n",
    "expr_ord = expr_ord.astype(float)\n",
    "y = np.log1p(np.expm1(expr_ord).sum(axis=1)).to_numpy()\n",
    "\n",
    "\n",
    "\n",
    "ax.set_ylabel('log1p(sum of normalized counts)')\n",
    "ax.plot(x, y, marker=\"o\", linestyle=\"-\", linewidth=1)\n",
    "ax.grid(True, alpha=0.3)\n",
    "\n",
    "# X ticks\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ordered_clusters, rotation=90)\n",
    "ax.set_xlabel(\"Clusters (grouped by subclass)\")\n",
    "\n",
    "# Legend: one entry per subclass band that actually appears\n",
    "legend_patches = []\n",
    "for sub, _, _ in spans:\n",
    "    if sub not in [p.get_label() for p in legend_patches]:\n",
    "        legend_patches.append(Patch(facecolor=subclass_colors.get(sub, default_color), alpha=0.4, label=sub))\n",
    "# Put legend above or to the side to avoid overlapping plots\n",
    "\n",
    "ax.legend(handles=legend_patches, loc=\"upper left\", bbox_to_anchor=(1.01, 1.0), frameon=False, title=\"Subclass bands\")\n",
    "ax.set_title('Total expression per cluster')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# If you also want to keep the explicit order for later use:\n",
    "ordered_clusters_list = ordered_clusters\n",
    "subclass_band_spans = spans  # [(subclass, start_idx, end_idx), ...]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5838cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sst_exp = panel_reference.X[:,panel_reference.var.index=='Sst'][panel_reference.obs.subclass_label=='Sst', :].toarray().flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d80082",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(sst_exp, bins = 100, log = True)\n",
    "plt.axvline(np.median(sst_exp), color = 'red')\n",
    "plt.text(np.median(sst_exp), 10, s = f'Median: {np.median(sst_exp):.1f}', color = 'red')\n",
    "plt.title('Sst expression in Sst cells')\n",
    "plt.xlabel('log1p(norm(counts))')\n",
    "plt.ylabel('Count (log scale)')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9057a201",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd5fa466",
   "metadata": {},
   "source": [
    "## Confusion matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16812e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "confusion_glia = pd.read_csv(\"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/evaluation_2023/forest_clfs/forest_clfs_adata1_spapros_10K_2023_taylored_glia_panel.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7acca85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "confusion_glia.index = confusion_glia['Unnamed: 0']\n",
    "confusion_glia = confusion_glia.drop(columns = ['Unnamed: 0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f1567eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.heatmap(\n",
    "    confusion_glia,\n",
    "    cmap=\"viridis\",     # or \"magma\", \"coolwarm\", etc.\n",
    "    annot=True,         # show values\n",
    "    fmt=\".3f\",          # 3 decimal places\n",
    "    cbar_kws={'label': 'Score'}\n",
    ")\n",
    "plt.title(\"Subclass similarity heatmap\")\n",
    "plt.ylabel(\"True subclass\")\n",
    "plt.xlabel(\"Predicted subclass\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfb37f28",
   "metadata": {},
   "source": [
    "## Testing expression outside our target subclasses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8095373d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading the yao2021 reference\n",
    "\n",
    "ALM_neuron_subclasses = [\n",
    "    \"Vip\",\n",
    "    \"Lamp5\",\n",
    "    \"Scng\",\n",
    "    \"Sst Chodl\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"L2/3 IT CTX\",\n",
    "    \"L4/5 IT CTX\",\n",
    "    \"L5 IT CTX\",\n",
    "    \"L6 IT CTX\",\n",
    "    \"L5 PT CTX\",\n",
    "    \"L4 RSP-ACA\",\n",
    "    \"L5/6 NP CTX\",\n",
    "    \"L6 CT CTX\",\n",
    "    \"L6b CTX\",\n",
    "]\n",
    "\n",
    "ALM_glia_subclasses = ['Oligo', 'Astro', 'SMC-Peri', 'VLMC', 'Micro-PVM', 'Endo', 'CR', 'Meis2']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "datapath = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/yao_2021')\n",
    "\n",
    "glia_reference = io.load_yao_2021_to_anndata(datapath, 'ALM', ALM_glia_subclasses)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1704c1a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c08bcd5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_genes(glia_reference, min_cells=1) # it crashes with a lot of genes expressed in no cells\n",
    "sc.pp.normalize_total(glia_reference)\n",
    "sc.pp.log1p(glia_reference)\n",
    "sc.pp.highly_variable_genes(glia_reference,flavor=\"cell_ranger\",n_top_genes=10000) #for a good panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ad48d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "savepath = \"/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/panels/spapros_10K_taylored_glia_2021\"\n",
    "\n",
    "prior_marker = [\n",
    "    \"Chodl\",\n",
    "    \"Cux2\",\n",
    "    \"Fezf2\",\n",
    "    \"Foxp2\",\n",
    "    \"Rorb\",\n",
    "    \"Pvalb\",\n",
    "    \"Lamp5\",\n",
    "    \"Adamts2\",\n",
    "    \"Slco2a1\",\n",
    "]\n",
    "\n",
    "selector = sp.se.ProbesetSelector(glia_reference, n=100, \n",
    "                                  celltype_key=\"subclass_label\", \n",
    "                                  verbosity=1, \n",
    "                                  n_jobs=-1, \n",
    "                                  save_dir=savepath, \n",
    "                                  preselected_genes=prior_marker)\n",
    "\n",
    "selected = selector.probeset[selector.probeset[\"selection\"]].copy()\n",
    "spapros_panel = list(selected.index)\n",
    "\n",
    "panel_reference = glia_reference[:,glia_reference.var_names.isin(spapros_panel)]\n",
    "\n",
    "#this is for the entropy constraint\n",
    "gene_priority = [\n",
    "    \"Chodl\",\n",
    "    \"Vip\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"Rorb\",\n",
    "    \"Lamp5\",\n",
    "    \"Cux2\",\n",
    "    \"Fezf2\",\n",
    "    \"Rrad\",\n",
    "    \"Adamts2\",\n",
    "    \"Foxp2\",\n",
    "    \"Npsr1\",\n",
    "    \"Slco2a1\",\n",
    "    \"Tshz2\",\n",
    "    \"Fam19a1\",\n",
    "    \"Calb1\",\n",
    "    \"Syt6\",\n",
    "    \"Blnk\",\n",
    "    \"Cplx3\",\n",
    "    \"Adarb2\",\n",
    "    \"Tcap\",\n",
    "    \"Rprm\",\n",
    "    \"Cpne9\",\n",
    "    \"Tac1\",\n",
    "    \"Ctgf\",\n",
    "    \"Erbb4\",\n",
    "    \"Slc30a3\",\n",
    "    \"Synpr\",\n",
    "    \"Cdh9\",\n",
    "    \"Tnnc1\",\n",
    "    \"Il1rapl2\",\n",
    "    \"Meis2\",\n",
    "    \"Crh\",\n",
    "    \"Adcyap1\",\n",
    "    \"Rgs12\",\n",
    "    \"Lypd1\",\n",
    "    \"Igfbp6\",\n",
    "    \"Sema5a\",\n",
    "    \"Cpne7\",\n",
    "    \"S100a6\",\n",
    "    \"Lypd6\",\n",
    "    \"Calb2\",\n",
    "    \"Pamr1\",\n",
    "    \"Cidea\",\n",
    "    \"Medag\",\n",
    "    \"Ptprk\",\n",
    "    \"Pcp4l1\",\n",
    "    \"Nxph1\",\n",
    "    \"Crym\",\n",
    "    \"Ptprm\",\n",
    "    \"Cbln4\",\n",
    "    \"Igfbp4\",\n",
    "    \"Pnoc\",\n",
    "    \"Penk\",\n",
    "    \"Npy\",\n",
    "    \"Egfr\",\n",
    "    \"Cryab\",\n",
    "    \"Grik1\",\n",
    "    \"Kctd8\",\n",
    "    \"Car4\",\n",
    "    \"S100b\",\n",
    "    \"Nr2f2\",\n",
    "    \"Pthlh\",\n",
    "    \"Prss23\",\n",
    "    \"Cdh20\",\n",
    "    \"Ptprz1\",\n",
    "    \"Vwc2l\",\n",
    "    \"Elfn1\",\n",
    "    \"Crhbp\",\n",
    "    \"Gad1\",\n",
    "    \"Angpt1\",\n",
    "    \"Adamts3\",\n",
    "    \"Nxph4\",\n",
    "    \"Npr3\",\n",
    "    \"Sema5b\",\n",
    "    \"Ddit4l\",\n",
    "    \"Kit\",\n",
    "    \"Serpine2\",\n",
    "    \"Unc13c\",\n",
    "    \"Cd63\",\n",
    "    \"Cartpt\",\n",
    "    \"Thsd7a\",\n",
    "    \"Tac2\",\n",
    "    \"Moxd1\",\n",
    "    \"Rmst\",\n",
    "    \"Sema3c\",\n",
    "    \"Rspo1\",\n",
    "    \"Sox6\",\n",
    "    \"Ppapdc1a\",\n",
    "    \"Lgals1\",\n",
    "    \"Ackr3\",\n",
    "    \"Timp3\",\n",
    "    \"Npas1\",\n",
    "    \"Reln\",\n",
    "    \"Vwc2\",\n",
    "    \"Pcdh8\",\n",
    "    \"Hgf\",\n",
    "    \"Pcdh20\",\n",
    "    \"Cp\",\n",
    "    \"Ndnf\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77a4b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_persubclass = ut.calculate_expression_per_taxa(panel_reference, 'subclass_label')\n",
    "\n",
    "\n",
    "maxmean = expression_persubclass.max() - expression_persubclass.mean()\n",
    "\n",
    "#calculate entropy of each gene\n",
    "gene_H = entropy(expression_persubclass.to_numpy(dtype = float), base = 2, axis = 0)\n",
    "\n",
    "h_thresh = 3.0        # entropy threshold (bits)\n",
    "mm_thresh = 2.0       # max-mean threshold (same units as your data)\n",
    "mask = (gene_H > h_thresh) & (maxmean < mm_thresh)\n",
    "almost_mask = (gene_H > h_thresh) ^ (maxmean < mm_thresh)\n",
    "colors_1 = np.where(mask, 1, 0)\n",
    "colors_2 = np.where(almost_mask, 2, 0)\n",
    "colors_sum = colors_1+colors_2\n",
    "colist = ['black', 'red', 'orange']\n",
    "colors = [colist[i] for i in colors_sum]\n",
    "\n",
    "plt.scatter(maxmean, gene_H, c=colors, s=20)\n",
    "plt.ylabel('per gene Shannon entropy (bits)')\n",
    "plt.xlabel('per gene max-mean (log1p(norm(counts))')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e2cf56",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "expr_ord = expression_persubclass\n",
    "# ---- Plot: one subplot (row) per gene; shaded subclass bands ----\n",
    "genes_order =np.argsort((expr_ord.max() - expr_ord.mean()).values)\n",
    "\n",
    "genes = list(expr_ord.columns[genes_order]) #sort genes according to max-mean quality metric\n",
    "\n",
    "plot_maxmean = maxmean[np.argsort(maxmean)]#sort metrics according to the same metric to preserve correspondence\n",
    "plot_entropy = gene_H[np.argsort(maxmean)]\n",
    "plot_colors = np.array(colors)[np.argsort(maxmean)]\n",
    "\n",
    "n_genes = len(genes)\n",
    "\n",
    "# ---- Layout knobs ----\n",
    "height_per_subplot = 4.4   # bump this up if still too squished\n",
    "fig_w = 12\n",
    "legend_right_margin = 0.80 # reserve space for legend (0.80 = 80% of width for plots)\n",
    "\n",
    "fig, axes = plt.subplots(\n",
    "    n_genes, 1,\n",
    "    figsize=(fig_w, max(4, height_per_subplot * n_genes)),\n",
    ")\n",
    "\n",
    "\n",
    "if n_genes == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "x = np.arange(len(glia_reference.obs['subclass_label'].unique()))\n",
    "\n",
    "for i, gene in enumerate(genes):\n",
    "    ax = axes[i]\n",
    "    # gene trace\n",
    "    y = expr_ord[gene].astype(float).to_numpy()\n",
    "    ax.plot(x, y, marker=\"o\", linestyle=\"none\", markersize=20, linewidth=1)\n",
    "    ax.text(0.99, 0.95,\n",
    "        f'Entropy: {plot_entropy[i]:.3f}\\nMax-mean: {plot_maxmean[i]:.3f}\\n{gene}',\n",
    "        transform=ax.transAxes,  # <--- relative to subplot axes\n",
    "        ha=\"right\", va=\"top\", \n",
    "        color = plot_colors[i])\n",
    "\n",
    "    ax.set_ylabel(f'log1p(norm({gene}))')\n",
    "    ax.set_xticks(np.arange(len(expr_ord.index)), labels=expr_ord.index)\n",
    "    ax.grid(True, alpha=0.3)\n",
    "    ax.set_ylim(-0.5, 10)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4f566a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 1) Total expression per subclass (undo log1p, sum, log1p again)\n",
    "expr_values = expression_persubclass.to_numpy(dtype=float)\n",
    "total_expr = np.log1p(np.expm1(expr_values).sum(axis=1))\n",
    "total_expr = pd.Series(total_expr, index=expression_persubclass.index, name=\"total_expr\")\n",
    "\n",
    "# 2) Cell counts per subclass\n",
    "cells_per_subclass = (\n",
    "    glia_reference.obs[\"subclass_label\"]\n",
    "    .value_counts()\n",
    "    .rename_axis(\"subclass\")\n",
    "    .to_frame(\"n_cells\")\n",
    "    .sort_index()\n",
    ")\n",
    "\n",
    "# 3) Align both by subclass index\n",
    "df = pd.concat([total_expr, cells_per_subclass[\"n_cells\"]], axis=1, join=\"inner\")\n",
    "\n",
    "# 4) Scatter plot\n",
    "fig, ax = plt.subplots(figsize=(7, 5))\n",
    "ax.scatter(df[\"n_cells\"], df[\"total_expr\"], s=60)\n",
    "\n",
    "# Add labels\n",
    "for subclass, row in df.iterrows():\n",
    "    ax.text(row[\"n_cells\"], row[\"total_expr\"], subclass,\n",
    "            fontsize=8, ha=\"center\", va=\"bottom\")\n",
    "\n",
    "ax.set_xlabel(\"Number of cells per subclass\")\n",
    "ax.set_ylabel(\"log1p(sum of normalized counts)\")\n",
    "ax.set_title(\"Total expression vs. cell count per subclass\")\n",
    "ax.grid(alpha=0.3)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c451c5da",
   "metadata": {},
   "outputs": [],
   "source": [
    "flatvalues = expression_persubclass.values.flatten()\n",
    "for value in flatvalues:\n",
    "    assert is_type(value, float), f'Negative value {value} found somewhere'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f69c46",
   "metadata": {},
   "source": [
    "Omit Meis2, CR, SMC Peri and Micro PVM, penalise for max expression among the rest, sigmoid at 7?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e4b83b",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 7\n",
    "\n",
    "\n",
    "ALM_glia_subclasses = ['Oligo', 'Astro', 'VLMC', 'Endo']\n",
    "datapath = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/yao_2021')\n",
    "glia_penalty_reference = io.load_yao_2021_to_anndata(datapath, 'ALM', ALM_glia_subclasses)\n",
    "\n",
    "sc.pp.filter_genes(glia_penalty_reference, min_cells=1) # it crashes with a lot of genes expressed in no cells\n",
    "sc.pp.normalize_total(glia_penalty_reference)\n",
    "sc.pp.log1p(glia_penalty_reference)\n",
    "sc.pp.highly_variable_genes(glia_penalty_reference,flavor=\"cell_ranger\",n_top_genes=10000) #for a good panel\n",
    "\n",
    "expression_persubclass = ut.calculate_expression_per_taxa(glia_penalty_reference, 'subclass_label')\n",
    "\n",
    "max_values = expression_persubclass.max()\n",
    "\n",
    "penalty = [1/(1+np.power(np.e, -k*(value-threshold))) for value in max_values]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97bd3ed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_values = expression_persubclass.max()\n",
    "\n",
    "k = 6\n",
    "threshold = 7\n",
    "\n",
    "penalty = [1/(1+np.power(np.e, -k*(value-threshold))) for value in max_values]\n",
    "\n",
    "plt.hist(penalty)\n",
    "plt.title(f'Glial expression penalty for k={k} and threshold={threshold}')\n",
    "plt.xlabel('Penalty')\n",
    "plt.ylabel('# genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e27a61c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_genes = np.array(expression_persubclass.columns)[np.argsort(penalty)]\n",
    "pd.DataFrame.from_dict({'gene': ordered_genes, 'penalty': np.sort(penalty)}).to_csv('glia_penalty.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b4ab378",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "809e0f29",
   "metadata": {},
   "source": [
    "## VIsp panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0947ec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALM_neuron_subclasses = [\n",
    "    \"Vip\",\n",
    "    \"Lamp5\",\n",
    "    \"Scng\",\n",
    "    \"Sst Chodl\",\n",
    "    \"Sst\",\n",
    "    \"Pvalb\",\n",
    "    \"L2/3 IT CTX\",\n",
    "    \"L4/5 IT CTX\",\n",
    "    \"L5 IT CTX\",\n",
    "    \"L6 IT CTX\",\n",
    "    \"L5 PT CTX\",\n",
    "    \"L4 RSP-ACA\",\n",
    "    \"L5/6 NP CTX\",\n",
    "    \"L6 CT CTX\",\n",
    "    \"L6b CTX\",\n",
    "]\n",
    "\n",
    "datapath = Path('/nemo/lab/znamenskiyp/home/shared/projects/colasa_MOs_panel/yao_2021')\n",
    "\n",
    "reference = io.load_yao_2021_to_anndata(datapath, 'VISp', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "837da742",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5d018cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get value counts of subclass_label\n",
    "subclass_counts = reference.obs['subclass_label'].value_counts()\n",
    "\n",
    "# Convert to DataFrame for better formatting\n",
    "df_counts = pd.DataFrame({\n",
    "    'Subclass': subclass_counts.index,\n",
    "    'Number of cells': subclass_counts.values\n",
    "})\n",
    "\n",
    "# Add total row\n",
    "total_cells = df_counts['Number of cells'].sum()\n",
    "df_counts.loc[len(df_counts)] = ['TOTAL', total_cells]\n",
    "\n",
    "print(df_counts.to_string(index=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "334c2260",
   "metadata": {},
   "source": [
    "L4/5 IT CTX\n",
    "Sst\n",
    "L6 IT CTX\n",
    "Vip\n",
    "Pvalb\n",
    "Lamp5\n",
    "L2/3 IT CTX\n",
    "L6 CT CTX\n",
    "L5 IT CTX\n",
    "L5 PT CTX\n",
    "L5/6 NP CTX\n",
    "L6b CTX\n",
    "Sst Chodl\n",
    "Car3\n",
    "Sncg\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spapros",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
